pkg/index/key_index.go

```diff
@@ -59,24 +60,41 @@ type keyIndex struct {
 	generations []generation
 }
 
-// todo: return error instead to panic
 // put puts a Revision to the keyIndex.
 func (ki *keyIndex) put(main int64, sub int64, nodes []string) {
 	rev := Revision{main: main, sub: sub, nodes: nodes}
 
-	if !rev.GreaterThan(ki.modified) {
-		panic(fmt.Errorf("store.keyindex: put with unexpected smaller Revision [%v / %v]", rev, ki.modified))
-	}
 	if len(ki.generations) == 0 {
 		ki.generations = append(ki.generations, generation{})
 	}
 	g := &ki.generations[len(ki.generations)-1]
 	if len(g.revs) == 0 { // create a new key
 		g.created = Revision{main: int64(inc.GetGlobalIncreasingRevision())}
 	}
-	g.revs = append(g.revs, rev)
-	g.ver++
-	ki.modified = rev
+
+	if !rev.GreaterThan(ki.modified) {
+		klog.Warningf("rev %v is older than the existing latest rev %v", rev, ki.modified)
+		// older than existent revs; locate its position in order
+		var i int
+		for i = len(g.revs) - 1; i >= 0; i-- {
```

Comment:
The revs are sorted. Not sure why u need to compare each of them.

Reply:
I remember the change is to adopt to multiple threads completing for updates. Therefore, if two threads are updating the the revision list. The origin one is [1, 2, 5]. The two threads get the same list. The one with rev 3 first update the list to [1,2,3,5]. Then the one with rev 4 know the position to insert is 2 and then the rev is changed to [1,2,4,3,5]?

Reply:
index should have lock to protect such corruption happen

Reply:
I got it. I found there was a RW lock. My another concern is whether we can write to the PAST. Considering we might have a index cluster for HA in the future. Do we need to sync the old versions to other nodes?